bson/streaming_value_reader.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (C) MongoDB, Inc. 2025-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package bson
+
+import (
+	"bufio"
+	"io"
+)
+
+// streamingValueReader reads from an ioReader wrapped in a bufio.Reader. It
+// first reads the BSON length header, then ensures it only ever reads exactly
+// that many bytes.
+//
+// Note: this approach trades memory usage for extra buffering and reader calls,
+// so it is less performanted than the in-memory bufferedValueReader.
+type streamingValueReader struct {
```

Comment:
Optional: This isn't a `ValueReader`, so the name is a bit confusing. Consider a name like `streamingByteSrc`.